### PR TITLE
Separate bacnet Ip connection and Serial connection.

### PR DESCRIPTION
**Pull request by hhou2011**
@Pacane

### CODE REVIEW

File: src/main/java/bacnet/BacnetConn.java [hunk 1]
```diff
@@ -354,6 +294,18 @@ public void run() {
...
+		return "10";
+	}
+
+	abstract Network createNetwork(int localNetworkNumber);
+
+	abstract void addTransportParameters(Action act);
+
+	abstract void setTransportAtrributions(ActionResult event);
+
+	abstract BacnetConn getBacnetConnection(BacnetLink link2, Node newnode);
```
Comment (Pacane) [inline src/main/java/bacnet/BacnetConn.java]: link2?

File: src/main/java/bacnet/BacnetConn.java [hunk 2]
```diff
@@ -511,53 +436,32 @@ void stop() {
...
-			int segwin = event.getParameter("segment window", ValueType.NUMBER).getNumber().intValue();
+			int segmentWindow = event.getParameter("segment window", ValueType.NUMBER).getNumber().intValue();
 			int retries = event.getParameter("retries", ValueType.NUMBER).getNumber().intValue();
-			int locdevId = event.getParameter("local device id", ValueType.NUMBER).getNumber().intValue();
-			String locdevName = event.getParameter("local device name", ValueType.STRING).getString();
-			String locdevVend = event.getParameter("local device vendor", ValueType.STRING).getString();
+			int localDeviceId = event.getParameter("local device id", ValueType.NUMBER).getNumber().intValue();
+			String localDeviceName = event.getParameter("local device name", ValueType.STRING).getString();
+			String localDeviceVendor = event.getParameter("local device vendor", ValueType.STRING).getString();
 			long interval = (long) (1000
```
Comment (Pacane) [inline src/main/java/bacnet/BacnetConn.java]: could be named `intervalInMilliseconds` or something like that.

File: src/main/java/bacnet/BacnetConn.java [hunk 3]
```diff
@@ -1044,6 +954,10 @@ public LocalDevice getLocalDevice() {
 		return ObjectToPoint;
 	}
 
+	void registerAsFeignDevice(Network newwork) {
```
Comment (Pacane) [inline src/main/java/bacnet/BacnetConn.java]: network

File: src/main/java/bacnet/BacnetConn.java [hunk 4]
```diff
@@ -1044,6 +954,10 @@ public LocalDevice getLocalDevice() {
 		return ObjectToPoint;
 	}
 
+	void registerAsFeignDevice(Network newwork) {
+
+	}
+
 	// // Not production code, for lisener's mockup test only
 	// public DeviceEventListener getListener() {
```
Comment (Pacane) [inline src/main/java/bacnet/BacnetConn.java]: remove this if it's not useful

File: src/main/java/bacnet/BacnetIpConnection.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+	Network createNetwork(int localNetworkNumber) {
+		IpNetworkBuilder builder = new IpNetworkBuilder();
+		Network network = builder.broadcastIp(broadcastIp).port(port).localBindAddress(localBindAddress)
+				.localNetworkNumber(localNetworkNumber).build();
+		return network;
+	}
+
+	private void parseBBMD(String bbmdIpList) {
```
Comment (Pacane) [inline src/main/java/bacnet/BacnetIpConnection.java]: what is BBMD? What's the acronym for?

Comment (Pacane) [reply src/main/java/bacnet/BacnetIpConnection.java]: - Isn't there already something to parse IP addresses?

Comment (Pacane) [reply src/main/java/bacnet/BacnetIpConnection.java]: Maybe something like this http://stackoverflow.com/questions/2345063/java-common-way-to-validate-and-convert-hostport-to-inetsocketaddress

File: src/main/java/bacnet/BacnetIpConnection.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	void registerAsFeignDevice(Network network) {
+		if (!isRegisteredAsForegnDevice)
+			return;
+
+		for (Map.Entry<String, Integer> entry : bbmdIpToPort.entrySet()) {
+			String bbmdIp = entry.getKey();
+			Integer bbmdPort = entry.getValue();
+			try {
+				((IpNetwork) network)
+						.registerAsForeignDevice(new InetSocketAddress(InetAddress.getByName(bbmdIp), bbmdPort), 100);
```
Comment (Pacane) [inline src/main/java/bacnet/BacnetIpConnection.java]: magic number

**Review by Pacane — CHANGES_REQUESTED** [review]

File: src/main/java/bacnet/BacnetConn.java (hunk 1 above)
Comment (hhou2011) [reply src/main/java/bacnet/BacnetConn.java]: Typo.

File: src/main/java/bacnet/BacnetIpConnection.java (hunk 5 above)
Comment (hhou2011) [reply src/main/java/bacnet/BacnetIpConnection.java]: bbmd = broadcast management device. Usually the short name is used in the bacnet protocol.

Comment (hhou2011) [reply src/main/java/bacnet/BacnetIpConnection.java]: Yes, I use pattern now.

**Review by Pacane — APPROVED** [review]

**Pull request merged**